cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java

```diff
@@ -583,87 +503,160 @@ private void emitConfigMetrics() {
+      this.maxEventTime = maxEventTime;
+    }
 
-    OffsetTime(long offset, long eventTime) {
+    @Override
+    public MutableKafkaOffset getOffset() {
+      return offset;
+    }
+
+    void setOffset(MutableKafkaOffset offset) {
       this.offset = offset;
-      this.eventTime = eventTime;
     }
 
     @Override
-    public int compareTo(OffsetTime o) {
-      return Long.compare(this.offset, o.offset);
+    public long getMaxEventTime() {
+      return maxEventTime;
     }
 
-    long getOffset() {
-      return offset;
+    void setMaxEventTime(long maxEventTime) {
+      this.maxEventTime = maxEventTime;
     }
 
-    long getEventTime() {
-      return eventTime;
+    @Override
+    public String toString() {
+      return "MutableCheckpoint{" +
+        "offset=" + offset +
+        ", maxEventTime=" + maxEventTime +
+        '}';
     }
   }
 
   /**
-   * A mutable implementation of {@link Checkpoint}.
+   * A mutable implementation of {@link KafkaOffset}.
    */
-  private static final class MutableCheckpoint extends Checkpoint {
-
+  private static final class MutableKafkaOffset extends KafkaOffset {
     private long nextOffset;
     private long nextEventTime;
-    private long maxEventTime;
 
-    MutableCheckpoint(Checkpoint other) {
-      this(other.getNextOffset(), other.getNextEventTime(), other.getMaxEventTime());
+    MutableKafkaOffset(KafkaOffset offset) {
+      this(offset.getNextOffset(), offset.getNextEventTime());
     }
 
-    MutableCheckpoint(long nextOffset, long nextEventTime, long maxEventTime) {
-      super(nextOffset, nextEventTime, maxEventTime);
-      this.nextOffset = nextOffset;
-      this.nextEventTime = nextEventTime;
-      this.maxEventTime = maxEventTime;
+    MutableKafkaOffset(long nextOffset, long nextEventTime) {
+      super(nextOffset, nextEventTime);
     }
 
     @Override
     public long getNextOffset() {
       return nextOffset;
     }
 
-    MutableCheckpoint setNextOffset(long nextOffset) {
+    void setNextOffset(long nextOffset) {
       this.nextOffset = nextOffset;
-      return this;
     }
 
     @Override
     public long getNextEventTime() {
       return nextEventTime;
     }
 
-    MutableCheckpoint setNextEvenTime(long nextEventTime) {
+    void setNextEventTime(long nextEventTime) {
       this.nextEventTime = nextEventTime;
-      return this;
     }
 
     @Override
-    public long getMaxEventTime() {
-      return maxEventTime;
+    public String toString() {
+      return "MutableKafkaOffset{" +
+        "nextOffset=" + nextOffset +
+        ", nextEventTime=" + nextEventTime +
+        '}';
     }
+  }
 
-    MutableCheckpoint setMaxEventTime(long maxEventTime) {
-      this.maxEventTime = maxEventTime;
-      return this;
+  /**
+   * Transforms kafka {@code MessageAndOffset} iterator to {@code ProcessorEvent} iterator.
+   */
+  private final class KafkaMessageTransformIterator implements Iterator<ProcessorEvent<KafkaOffset>> {
```

Comment:
extends from `AbstractIterator` instead of implementing directly. Alternatively, you can use Java stream and map.